### PR TITLE
Client library docs

### LINKED ISSUE #1: [CONTENT] Client library reference
A page should be created that show the core methods in some detail and describe the parameters, return values, etc.

### PR DESCRIPTION
**Pull request by johndmulhausen**
Resolves #1 

Covers the current methods from https://github.com/web3-storage/web3.storage/tree/main/packages/client

TODO: `status()` coverage does not include good details about the response yet as I couldn't find it in the source code.

For discussion: How do we like this format?

### CODE REVIEW

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- This phrase is duplicated. You should probably leave only “Store Files”.
  ```
  ...irst obtain an API key. # Constructor # Store Files Store files using the pu...
                                             ^^^^^^^^^^^^^^^^^^^^^^^
  ```
- This phrase is duplicated. You should probably leave only “Retrieve Files”.
  ```
  ...D of the uploaded CAR file. # Example # Retrieve Files Retrieve files using ...
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ... Which to use?: In the vast majority of cases you will want to use the files...
                                             ^^^^^
  ```
- As an alternative to the over-used intensifier ‘very’, consider replacing this phrase.
  ```
  ...he exception is when you are retrieving very large files in a Node.js contex...
                                             ^^^^^^^^^^
  ```
- This phrase is duplicated. You should probably leave only “Retrieve metadata”.
  ```
  ...onsume UnixFSEntry objects. # Example # Retrieve metadata Retrieve metadata ...
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Consider using either the past participle “packed” or the present participle “packing” here.
  ```
  ...he client library will be automatically pack your uploads into a content add...
                                             ^^^^
  ```
- This phrase is duplicated. You should probably leave only “Store Files”.
  ```
  ...D to retrieve the file. # Constructor # Store Files Store files using the pu...
                                             ^^^^^^^^^^^^^^^^^^^^^^^
  ```
- This phrase is duplicated. You should probably leave only “Retrieve Files”.
  ```
  ...D of the uploaded CAR file. # Example # Retrieve Files Retrieve files using ...
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- This phrase is duplicated. You should probably leave only “Retrieve metadata”.
  ```
  ...onsume UnixFSEntry objects. # Example # Retrieve metadata Retrieve metadata ...
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- It appears that a comma is missing here.
  ```
  ... web3.storage network. TODO: As of this writing the structure of the Metadat...
                                             ^^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Consider using either the past participle “packed” or the present participle “packing” here.
  ```
  ...he client library will be automatically pack your uploads into a content add...
                                             ^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- It appears that a comma is missing here.
  ```
  ... web3.storage network. TODO: As of this writing the structure of the Metadat...
                                             ^^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

File: docs/reference/client-library.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+description: Integrate web3.storage into your code using a client library for your programming language.
+---
+
+# JavaScript
+
+## Get started
+
+To use the JavaScript client library for web3.storage, you must first [obtain an API key](../how-tos/regenerate-api-key.md).
+
+The client library will automatically pack your uploads into a content addressible archive (CAR) file for uploading to the web3.storage service, which stores data as blocks prefixed with the content ID (CID) derived from the hash of the data, in a file with a `.car` extension. You can then use the CID to retrieve the file.
```
Comment (olizilla) [inline docs/reference/client-library.md:12]: This is accurate, tho strictly it doesn't put it into a file with a `.car` extention, as by that point it's just dealing with a stream of bytes. 

We could frame this more around the client deriving the content address for your files *before* uploading them so we dont have to trust the service, verifiabilty is an intrinsic part of the process.

File: docs/reference/client-library.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
+---
+title: JavaScript Client Library
+description: Integrate web3.storage into your code using a client library for your programming language.
```
Comment (johnnymatthews) [inline docs/reference/client-library.md:3]:
```suggestion
description: Integrate Web3.Storage into your code using a client library for your programming language.
```

Comment (johnnymatthews) [reply docs/reference/client-library.md:3]: I remember someone saying we should capitalize Web3.Storage.

File: docs/reference/client-library.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+---
+
+# JavaScript
+
+## Get started
+
+To use the JavaScript client library for web3.storage, you must first [obtain an API key](../how-tos/regenerate-api-key.md).
```
Comment (johnnymatthews) [inline docs/reference/client-library.md:10]:
```suggestion
To use the JavaScript client library for Web3.Storage, you must first [obtain an API key](../how-tos/regenerate-api-key.md).
```

File: docs/reference/client-library.md (hunk 1 above)
Comment (johnnymatthews) [reply docs/reference/client-library.md:12]:
```suggestion
The client library will automatically pack your uploads into a content addressible archive (CAR) file for uploading to the Web3.Storage service, which stores data as blocks prefixed with the content ID (CID) derived from the hash of the data. You can then use the CID to retrieve the file.
```

File: docs/reference/client-library.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+Parameters are supplied in positional order.
+
+| Number | Type | Description |
+| ------ | ---- | ----------- |
+| 1 | `string` | A string containing the CID of the CAR file to be retrieved. |
+
+### Return value
+
+The method returns a `Web3Response` object, which extends the [Fetch API response object](https://developer.mozilla.org/en-US/docs/Web/API/Response) to add two iterator methods unique to the web3.storage client library: `files()` and `unixFsIterator()`.
```
Comment (johnnymatthews) [inline docs/reference/client-library.md:74]:
```suggestion
The method returns a `Web3Response` object, which extends the [Fetch API response object](https://developer.mozilla.org/en-US/docs/Web/API/Response) to add two iterator methods unique to the Web3.Storage client library: `files()` and `unixFsIterator()`.
```

File: docs/reference/client-library.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+The `Web3File` type extends [the generic JavaScript `File` type](https://developer.mozilla.org/en-US/docs/Web/API/File), adding a `string` property for the CID of the given file named `cid`, as shown in the example below. This will be different from the CID of the CAR that contains the file, which you specified when calling `get()`.
+
+> **Note**: In addition to the `files()` method, you can also use the `unixFsIterator()` method, which will return your requested files as a  `AsyncIterable<UnixFSEntry>` object, which is an iterable collection of [`UnixFSEntry`](https://github.com/ipfs/js-ipfs-unixfs/blob/master/packages/ipfs-unixfs-exporter/README.md#unixfsentry) objects. Each object represents a file that was uploaded in the CAR file with the supplied CID. However, in the vast majority of cases, you will want to use the `files()` method. The exception is when you are retrieving files in a Node.js context, and you have existing IPFS tooling that knows how to consume `UnixFSEntry` objects.
```
Comment (johnnymatthews) [inline docs/reference/client-library.md:80]:
```suggestion
:::tip Using UnixFSEntry
In addition to the `files()` method, you can also use the `unixFsIterator()` method, which will return your requested files as a  `AsyncIterable<UnixFSEntry>` object, which is an iterable collection of [`UnixFSEntry`](https://github.com/ipfs/js-ipfs-unixfs/blob/master/packages/ipfs-unixfs-exporter/README.md#unixfsentry) objects. Each object represents a file that was uploaded in the CAR file with the supplied CID. However, in the vast majority of cases, you will want to use the `files()` method. The exception is when you are retrieving files in a Node.js context, and you have existing IPFS tooling that knows how to consume `UnixFSEntry` objects.
:::
```

**Review by johnnymatthews — CHANGES_REQUESTED** [review]

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by olizilla** [comment]
Of note, i think we settled on calling it `Web3 Storage` in prose and `web3.storage` when referring to the url or the client library. Never caps&dots together `Web3.Storage`

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/how-tos/retrieve.md`:
- The following link returned HTTP status code 301 (Moved Permanently):
  https://dweb.link/ipfs/bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu
- The following link returned HTTP status code 301 (Moved Permanently):
  https://dweb.link/ipfs/bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu/not-distributed.jpg
- Possibly missing comma found.
  ```
  ...t Identifier (CID)  (opens new window). First you'll need to create a Web3.S...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... you have a client, you can call client.get, passing in a CID string:  The g...
                                             ^^^
  ```
- An apostrophe may be missing.
  ```
  ...access to the retrieved IPFS data.  The files method returns an array of Web...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ...is guide, we'll use the gateway at dweb.link, but you can check the list of ...
                                             ^^^^
  ```



### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...iles being uploaded. See below for more details An {options} object is a JSO...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... the iterator represent files. If entry.type == 'directory', the entry repre...
                                             ^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by terichadbourne** [comment]
@olizilla I thought we settled on Web3.Storage to match how NFT.Storage is being referred to elsewhere. Lemme know if you have a more recent verdict than the one at https://protocollabs.slack.com/archives/C01MPC37Z37/p1625154646106100?thread_ts=1625154546.104600&cid=C01MPC37Z37

**Comment by terichadbourne** [comment]
Just confirmed with @jnthnvctr that this is still meant to be Web3.Storage (w/ dot, w/o space).

**Comment by terichadbourne** [comment]
Note that the Tutorial section in the left nav was intentionally removed elsewhere because it's not a strict M1 goal to get that done, but it appears in this preview. I tried removing it by deleting the `docs/tutorials` folder but that only got rid of the contents and not the Tutorials header in left nav. @johndmulhausen Mind reminding me where else I need to go to remove it?

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/how-tos/retrieve.md`:
- The following link returned HTTP status code 301 (Moved Permanently):
  https://dweb.link/ipfs/bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu
- The following link returned HTTP status code 301 (Moved Permanently):
  https://dweb.link/ipfs/bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu/not-distributed.jpg
- Possibly missing comma found.
  ```
  ...t Identifier (CID)  (opens new window). First you'll need to create a Web3.S...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... you have a client, you can call client.get, passing in a CID string:  The g...
                                             ^^^
  ```
- An apostrophe may be missing.
  ```
  ...access to the retrieved IPFS data.  The files method returns an array of Web...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ...is guide, we'll use the gateway at dweb.link, but you can check the list of ...
                                             ^^^^
  ```



### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...iles being uploaded. See below for more details An {options} object is a JSO...
                                             ^^^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... the iterator represent files. If entry.type == 'directory', the entry repre...
                                             ^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/how-tos/retrieve.md`:
- The following link returned HTTP status code 301 (Moved Permanently):
  https://dweb.link/ipfs/bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu
- The following link returned HTTP status code 301 (Moved Permanently):
  https://dweb.link/ipfs/bafybeidd2gyhagleh47qeg77xqndy2qy3yzn4vkxmk775bg2t5lpuy7pcu/not-distributed.jpg
- Possibly missing comma found.
  ```
  ...t Identifier (CID)  (opens new window). First you'll need to create a Web3.S...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... you have a client, you can call client.get, passing in a CID string:  The g...
                                             ^^^
  ```
- An apostrophe may be missing.
  ```
  ...access to the retrieved IPFS data.  The files method returns an array of Web...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ...is guide, we'll use the gateway at dweb.link, but you can check the list of ...
                                             ^^^^
  ```



### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...iles being uploaded. See below for more details An {options} object is a JSO...
                                             ^^^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... the iterator represent files. If entry.type == 'directory', the entry repre...
                                             ^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/how-tos/retrieve.md`:
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... you have a client, you can call client.get, passing in a CID string:  The g...
                                             ^^^
  ```
- An apostrophe may be missing.
  ```
  ...access to the retrieved IPFS data.  The files method returns an array of Web...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ...is guide, we'll use the gateway at dweb.link, but you can check the list of ...
                                             ^^^^
  ```



### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... the iterator represent files. If entry.type == 'directory', the entry repre...
                                             ^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/how-tos/retrieve.md`:
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... you have a client, you can call client.get, passing in a CID string:  The g...
                                             ^^^
  ```
- An apostrophe may be missing.
  ```
  ...access to the retrieved IPFS data.  The files method returns an array of Web...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ...is guide, we'll use the gateway at dweb.link, but you can check the list of ...
                                             ^^^^
  ```



### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... the iterator represent files. If entry.type == 'directory', the entry repre...
                                             ^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Review by yusefnapora — APPROVED** [review]
This is great, thanks @johndmulhausen 👍 

I went ahead and pushed a few commits to fix up links and things, but I got a bit carried away and snuck in some edits to the storage howto (removing the references to web3-file). Hope you don't mind me cluttering up your PR :)

**Comment by filecorgi** [comment]
- Image optimization came back clean!
- Vuepress build was successful!

### `docs/how-tos/retrieve.md`:
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... you have a client, you can call client.get, passing in a CID string:  The g...
                                             ^^^
  ```
- An apostrophe may be missing.
  ```
  ...access to the retrieved IPFS data.  The files method returns an array of Web...
                                             ^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ...is guide, we'll use the gateway at dweb.link, but you can check the list of ...
                                             ^^^^
  ```
### `docs/how-tos/store.md`:
- The following link failed to connect:
  https://web3.storage/tokens/
- Only proper nouns start with an uppercase character (there are exceptions for headlines).
  ```
  ...APIs like fetch  (opens new window) and File  (opens new window). This guide...
                                             ^^^^
  ```
- Only proper nouns start with an uppercase character (there are exceptions for headlines).
  ```
  ...ile objects manually: Unique file names Try to give each file a unique file ...
                                             ^^^
  ```
- Apostrophe might be missing. Did you mean “file’s” (curly possessive apostrophe) or “file's” (straight possessive apostrophe)? (Some terms such as “farmers’ market” have a possessive apostrophe after the first noun.)
  ```
  ...one content archive, and linking to the files inside is much simpler if each...
                                             ^^^^^
  ```
- Consider using a more formal alternative to avoid the overused word “more”.
  ```
  ...the IPFS command line. You can also get more information about the status of...
                                             ^^^^^^^^^^^^^^^^
  ```


### `docs/reference/client-library.md`:
- Possibly missing preposition found.
  ```
  ... the JavaScript client library for Web3.Storage, you must first obtain an AP...
                                             ^^^^^^^
  ```
- If a new sentence starts here, add a space and start with an uppercase letter.
  ```
  ... the iterator represent files. If entry.type == 'directory', the entry repre...
                                             ^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```
- Possibly missing comma found.
  ```
  ...tional order. Number Type Description 1 string A string containing the CID o...
                                             ^^^^^^
  ```

**Pull request merged**